### PR TITLE
Include send-data missing headers and organize telemetry config variables

### PR DESCRIPTION
**Pull request by iunanua**
### What does this PR do?
- Include `dd-client-library-language` and `dd-client-library-version` headers
- Move `DD_TELEMETRY_HEARTBEAT_INTERVAL` to `config.js`

### Motivation
Some system-tests will check them

### Plugin Checklist

- [ ] Unit tests.
- [ ] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [ ] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [ ] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/documentation/blob/master/content/en/tracing/trace_collection/library_config/nodejs.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### CODE REVIEW

File: packages/dd-trace/src/telemetry/send-data.js [hunk 1]
```diff
@@ -1,4 +1,26 @@
 const request = require('../exporters/common/request')
+const { isTrue } = require('../util')
+
+const debug = process.env.DD_TELEMETRY_DEBUG
```
Comment (simon-id) [inline packages/dd-trace/src/telemetry/send-data.js:4]: this should be done in config.js instead no ?

Comment (iunanua) [reply packages/dd-trace/src/telemetry/send-data.js:4]: It is a internal flag to enable debugging requests, that's the reason I left it here. But I can move it to config.js.

Ah, [telemetry/index.js](https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/telemetry/index.js#L9) reads `DD_TELEMETRY_HEARTBEAT_INTERVAL` property in the same way (in the file itself instead of in config.js)...

Comment (simon-id) [reply packages/dd-trace/src/telemetry/send-data.js:4]: Ah, well idk then, ask the tracer team

Comment (iunanua) [reply packages/dd-trace/src/telemetry/send-data.js:4]: I could move to config.js both...

Comment (iunanua) [reply packages/dd-trace/src/telemetry/send-data.js]: moved both!

**Review by Qard — APPROVED** [review]

File: packages/dd-trace/src/config.js [hunk 2]
```diff
@@ -200,8 +200,11 @@ class Config {
       process.env.DD_TRACE_TELEMETRY_ENABLED,
       !process.env.AWS_LAMBDA_FUNCTION_NAME
     )
-    const DD_TELEMETRY_DEBUG_ENABLED = coalesce(
-      process.env.DD_TELEMETRY_DEBUG_ENABLED,
+    const DD_TELEMETRY_HEARTBEAT_INTERVAL = process.env.DD_TELEMETRY_HEARTBEAT_INTERVAL
+      ? parseInt(process.env.DD_TELEMETRY_HEARTBEAT_INTERVAL) * 1000
+      : 60000
+    const DD_TELEMETRY_DEBUG = coalesce(
```
Comment (rochdev) [inline packages/dd-trace/src/config.js:206]: Why not keep both names for backward compatibility?

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/dd-trace/src/config.js (hunk 2 above)
Comment (iunanua) [reply packages/dd-trace/src/config.js:206]: It is an IAST property and nobody is using it yet

**Review by rochdev — APPROVED** [review]

File: packages/dd-trace/src/telemetry/send-data.js [hunk 3]
```diff
@@ -1,4 +1,21 @@
 const request = require('../exporters/common/request')
+
+function getHeaders (reqType, debug, application) {
+  const headers = {
+    'Content-Type': 'application/json',
```
Comment (uurien) [inline packages/dd-trace/src/telemetry/send-data.js:5]: You have changed some keys chars from lowercase to uppercase, `Content-Type`, `DD-Telemetry-API-Version`...
Is this ok?

Comment (iunanua) [reply packages/dd-trace/src/telemetry/send-data.js:5]: Yes, the specification states that the names should be in camelcase and some system-tests check it

File: packages/dd-trace/test/config.spec.js [hunk 4]
```diff
@@ -786,10 +786,23 @@ describe('Config', () => {
 
     expect(config.telemetry).to.not.be.undefined
     expect(config.telemetry.enabled).to.be.true
+    expect(config.telemetry.heartbeatInterval).to.be.eq(60000)
```
Comment (simon-id) [inline packages/dd-trace/test/config.spec.js:789]: `to.be.eq...` sounds weird IMO, the whole point of these asserting libs is that it sounds like normal english.
If you look at the rest of the repo:
without be 1000+ results
image
with be 100+ results
image
There is a full order of magnitude difference, you even use `expect(options).to.deep.equal({` without the `.be` in a file below.
IMO the `eq`/`equal` keyword has the same grammar function as the `be` keyword.
Up to you to decide with the elements above.

File: packages/dd-trace/src/telemetry/send-data.js (hunk 3 above)
Comment (uurien) [reply packages/dd-trace/src/telemetry/send-data.js:5]: Hmm, but CamelCase is not like: `ContentType`, `DDClientLibraryLanguage` without `-`?

File: packages/dd-trace/test/telemetry/send-data.spec.js [hunk 5]
```diff
@@ -12,40 +17,78 @@ describe('sendData', () => {
       '../exporters/common/request': request
     })
   })
+
   it('should call to request (TCP)', () => {
-    sendDataModule.sendData({ hostname: '', port: '12345', tags: { 'runtime-id': '123' } }, 'test', 'test', 'req-type')
+    sendDataModule.sendData({ hostname: '', port: '12345', tags: { 'runtime-id': '123' } },
+      application, 'test', 'req-type')
```
Comment (simon-id) [inline packages/dd-trace/test/telemetry/send-data.spec.js:23]: very weird line break, this is more js idiomatic:
```suggestion
    sendDataModule.sendData({
      hostname: '',
      port: '12345',
      tags: { 'runtime-id': '123' }
    }, application, 'test', 'req-type')
```

File: packages/dd-trace/test/telemetry/send-data.spec.js [hunk 6]
```diff
@@ -12,40 +17,78 @@ describe('sendData', () => {
...
       url: 'unix:/foo/bar/baz',
       hostname: undefined,
       port: undefined
     })
   })
+
+  it('should add debug header if DD_TELEMETRY_DEBUG is present', () => {
+    sendDataModule = proxyquire('../../src/telemetry/send-data', {
+      '../exporters/common/request': request
+    })
```
Comment (simon-id) [inline packages/dd-trace/test/telemetry/send-data.spec.js:68]: isn't that already done in the beforeach ?

File: packages/dd-trace/src/config.js [hunk 7]
```diff
@@ -200,8 +200,11 @@ class Config {
       process.env.DD_TRACE_TELEMETRY_ENABLED,
       !process.env.AWS_LAMBDA_FUNCTION_NAME
     )
-    const DD_TELEMETRY_DEBUG_ENABLED = coalesce(
-      process.env.DD_TELEMETRY_DEBUG_ENABLED,
+    const DD_TELEMETRY_HEARTBEAT_INTERVAL = process.env.DD_TELEMETRY_HEARTBEAT_INTERVAL
+      ? parseInt(process.env.DD_TELEMETRY_HEARTBEAT_INTERVAL) * 1000
```
[the remark this answers is not in the document]
Comment (iunanua) [reply packages/dd-trace/src/config.js:204]: I guess it is very rare to want to set a decimal interval here.

File: packages/dd-trace/test/telemetry/send-data.spec.js (hunk 6 above)
Comment (iunanua) [reply packages/dd-trace/test/telemetry/send-data.spec.js:68]: true

File: packages/dd-trace/test/config.spec.js (hunk 4 above)
Comment (iunanua) [reply packages/dd-trace/test/config.spec.js:789]: changed to `to.eq`

File: packages/dd-trace/src/telemetry/index.js [hunk 8]
```diff
@@ -119,11 +116,13 @@ function start (aConfig, thePluginManager) {
   pluginManager = thePluginManager
   application = createAppObject()
   host = createHostObject()
+  heartbeatInterval = config.telemetry.heartbeatInterval || 60000
```
[the remark this answers is not in the document]
Comment (iunanua) [reply packages/dd-trace/src/telemetry/index.js:119]: true

File: packages/dd-trace/src/telemetry/send-data.js (hunk 3 above)
Comment (iunanua) [reply packages/dd-trace/src/telemetry/send-data.js:5]: CamelCaseHyphenSeparated :joy:

Comment (rochdev) [reply packages/dd-trace/src/telemetry/send-data.js:5]: Headers are case-insensitive, I don't understand why we're checking the casing anywhere.

Comment (Qard) [reply packages/dd-trace/src/telemetry/send-data.js:5]: Node.js also automatically lower-cases headers in a bunch of places.

Comment (iunanua) [reply packages/dd-trace/src/telemetry/send-data.js:5]: ok, system-tests changed to be case-insensitive https://github.com/DataDog/system-tests/pull/1111/commits/5230d8f350ec21fa63ea8b57785990e905d8dfa3

Comment (iunanua) [reply packages/dd-trace/src/telemetry/send-data.js]: and restored [original header capitalization ](https://github.com/DataDog/dd-trace-js/pull/3055/files#diff-d42dccc371917e1e5b6b4961ffd59ac8e264c4c6a8c650b940f388ebbdf009c3R5)

File: packages/dd-trace/test/telemetry/send-data.spec.js (hunk 5 above)
Comment (simon-id) [reply packages/dd-trace/test/telemetry/send-data.spec.js]: @iunanua you didn't update it correctly, look at the suggestion above ^

**Review by simon-id — APPROVED** [review]
LGTM but maybe get an approval from the tracing team too i guess ? :+1:

File: packages/dd-trace/src/appsec/iast/telemetry/logs.js [hunk 9]
```diff
@@ -28,7 +28,7 @@ function sendLogs () {
 }
 
 function isLevelEnabled (level) {
-  return isLogCollectionEnabled(config) && (level !== 'DEBUG' || config.telemetry.debug)
+  return isLogCollectionEnabled(config) && (level !== 'DEBUG' || config.telemetry.diagnosticLogCollector)
```
[the remark this answers is not in the document]
Comment (rochdev) [reply packages/dd-trace/src/appsec/iast/telemetry/logs.js:31]: Do we have a test for this? Even with the incorrect value it doesn't seem any test was failing.

File: packages/dd-trace/src/telemetry/index.js [hunk 10]
```diff
@@ -6,10 +6,6 @@ const os = require('os')
 const dependencies = require('./dependencies')
 const { sendData } = require('./send-data')
 
-const HEARTBEAT_INTERVAL = process.env.DD_TELEMETRY_HEARTBEAT_INTERVAL
```
Comment (rochdev) [inline packages/dd-trace/src/telemetry/index.js:9]: Why was this moved? It doesn't seem to support programmatic config or be used anywhere else.

File: packages/dd-trace/src/config.js [hunk 11]
```diff
@@ -375,6 +378,10 @@ ken|consumer_?(?:id|key|secret)|sign(?:ed|ature)?|auth(?:entication|orization)?)
       process.env.DD_TELEMETRY_LOG_COLLECTION_ENABLED,
       DD_IAST_ENABLED
     )
+    const DD_TELEMETRY_DIAGNOSTIC_LOG_COLLECTION_ENABLED = coalesce(
```
Comment (rochdev) [inline packages/dd-trace/src/config.js:381]: Why is this no longer enabled by default? How will we know if there is an issue?

File: packages/dd-trace/src/telemetry/send-data.js [hunk 12]
```diff
@@ -1,4 +1,21 @@
...
+function getHeaders (reqType, debug, application) {
+  const headers = {
+    'content-type': 'application/json',
+    'dd-telemetry-api-version': 'v1',
+    'dd-telemetry-request-type': reqType
+  }
+  if (debug) {
+    headers['dd-telemetry-debug-enabled'] = 'true'
+  }
+  if (application) {
```
Comment (rochdev) [inline packages/dd-trace/src/telemetry/send-data.js:12]: How could this be missing? These are 2 process-level information that should always be available.

File: packages/dd-trace/src/config.js (hunk 11 above)
Comment (rochdev) [reply packages/dd-trace/src/config.js:381]: Also, using a different term here could be confusing. Why "diagnostic" instead of "debug" like what is used elsewhere?

File: packages/dd-trace/src/config.js (hunk 2 above)
Comment (rochdev) [reply packages/dd-trace/src/config.js:206]: Is this even used anywhere at this point?

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/dd-trace/src/telemetry/index.js (hunk 10 above)
Comment (iunanua) [reply packages/dd-trace/src/telemetry/index.js:9]: I moved it to group all the telemetry properties reading in one place.

File: packages/dd-trace/src/telemetry/send-data.js (hunk 12 above)
Comment (iunanua) [reply packages/dd-trace/src/telemetry/send-data.js:12]: I guess they are added by the agent, so there is no need to add them in the tracer.

Comment (iunanua) [reply packages/dd-trace/src/telemetry/send-data.js]: I'm not sure :S, let me check it again.

Comment (iunanua) [reply packages/dd-trace/src/telemetry/send-data.js:12]: No, they are not added by the agent.

And they are mandatory headers according to the spec

I know the application data is included among the request data but it seems that the headers help to identify the request even if the backend fails to parse the request payload. For example to identify a particular library version and language, sending requests that cannot be parsed.

So it is useful for the backend that the tracer includes them.

File: packages/dd-trace/src/config.js (hunk 11 above)
Comment (rochdev) [reply packages/dd-trace/src/config.js:381]: This differs a lot from our current approach. Why not use `DEBUG` and `LOG_LEVEL` to be consistent? We'll probably want to normalize all of this properly across the library later on too.

File: packages/dd-trace/src/telemetry/send-data.js (hunk 12 above)
Comment (rochdev) [reply packages/dd-trace/src/telemetry/send-data.js:12]: It's fine if they're needed, but in which case would the condition be false? There would always be an `application`, and this it seems that this would always be true, so why the `if`?

Comment (iunanua) [reply packages/dd-trace/src/telemetry/send-data.js]: Because `sendData` method doesn't guarantee that `application` will always have a value and `getHeaders` method accesses some `application` properties. 
I felt safer with the `if`.
Removed

File: packages/dd-trace/src/config.js (hunk 11 above)
Comment (iunanua) [reply packages/dd-trace/src/config.js:381]: I'm going to remove `DD_TELEMETRY_DIAGNOSTIC_LOG_COLLECTION_ENABLED` for now because it has no effect. And I'll try to talk with the RFC creators.

Comment (iunanua) [reply packages/dd-trace/src/config.js:381]: Ok, we've decided to remove it until we are interested in collecting DEBUG or INFO traces.

But `DD_TELEMETRY_DIAGNOSTIC_LOG_COLLECTION_ENABLED` has other implications besides changing the log level. If `true`:
- logs with stack traces sent via telemetry are not redacted and therefore no sensitive information is removed from them.
- and it is supposed to be valid for 4 hours (to make sure that customers don’t forget)

Comment (simon-id) [reply packages/dd-trace/src/config.js]: that's super weird, and I agree with roch :/

Comment (iunanua) [reply packages/dd-trace/src/config.js]: it has been removed from our implementation but I suppose java and other teams are using it.
wdyt @uurien @rochdev?

Comment (rochdev) [reply packages/dd-trace/src/config.js]: Yeah let's make sure it makes sense _before_ implementing it.

File: packages/dd-trace/src/telemetry/send-data.js [hunk 13]
```diff
@@ -1,4 +1,20 @@
 const request = require('../exporters/common/request')
+
+function getHeaders (config, application, reqType) {
+  const headers = {
+    'content-type': 'application/json',
+    'dd-telemetry-api-version': 'v1',
+    'dd-telemetry-request-type': reqType,
+    'dd-client-library-language': application.language_name,
```
Comment (rochdev) [inline packages/dd-trace/src/telemetry/send-data.js:8]: Why are we repeating this data that is always available as other headers?

Comment (iunanua) [reply packages/dd-trace/src/telemetry/send-data.js:8]: `application` object, which contains language and version, is included in the request payload but apparently `dd-client-library-language` and `dd-client-library-version` headers are mandatory according to the spec. 

I gave you a reason in this [comment](https://github.com/DataDog/dd-trace-js/pull/3055#discussion_r1185954314)

**Review by rochdev — APPROVED** [review]
